src/autopas/tuning/tuningStrategy/fuzzyTuning/FuzzyTuning.cpp

```diff
@@ -0,0 +1,282 @@
+
+  // get the best configurations and everything that is within 10% of the best
+  const double THRESHOLD = 0.1;
+
+  const auto bestSuitability = configSuitabilities.front().second;
+  std::vector<ConfigurationPattern> bestConfigurations;
+  for (const auto &[pattern, suitability] : configSuitabilities) {
+    if (suitability < bestSuitability * (1 - THRESHOLD)) {
+      break;
+    }
+
+    AutoPasLog(DEBUG,
+               "FuzzyTuning: Adding configuration pattern {} with suitability {} since its within {}% of the "
+               "best suitability {}",
+               pattern.toString(), suitability, THRESHOLD * 100, bestSuitability);
+    bestConfigurations.push_back(pattern);
+  }
+
+  // 4. Only keep configurations that are in the best configurations
+  newSearchSpace.remove_if([&bestConfigurations](const Configuration &configuration) {
+    return std::none_of(bestConfigurations.begin(), bestConfigurations.end(),
+                        [&configuration](const auto &pattern) { return pattern.matches(configuration); });
+  });
+
+  AutoPasLog(DEBUG, "Fuzzy tuning selected {} out of {} possible configurations", newSearchSpace.size(),
+             configQueue.size());
+
+  configQueue.clear();
+  std::copy(newSearchSpace.rbegin(), newSearchSpace.rend(), std::back_inserter(configQueue));
+}
+
+std::tuple<std::shared_ptr<FuzzyControlSettings>, std::vector<std::shared_ptr<LinguisticVariable>>,
+           std::map<std::string, std::shared_ptr<OutputMapper>>,
+           std::map<std::string, std::shared_ptr<FuzzyControlSystem>>>
+FuzzyTuning::parse(const std::string &fuzzyRuleFilename) {
+  using namespace antlr4;
+  using namespace autopas::fuzzy_logic;
+
+  try {
+    // Read file
+    std::ifstream stream(fuzzyRuleFilename);
+    ANTLRInputStream input(stream);
+    FuzzyLanguageLexer lexer(&input);
+
+    // Lexer
+    FuzzyRuleErrorListener ErrorListener;
+    lexer.removeErrorListeners();
+    lexer.addErrorListener(&ErrorListener);
+
+    CommonTokenStream tokens(&lexer);
+
+    // Parser
+    FuzzyLanguageParser parser(&tokens);
+    parser.removeErrorListeners();
+    parser.addErrorListener(&ErrorListener);
+
+    tree::ParseTree *tree = parser.rule_file();
+
+    // Translation
+    TranslationVisitor visitor;
+    auto fuzzy_rule_program =
+        visitor.visit(tree)
+            .as<std::tuple<std::shared_ptr<FuzzyControlSettings>, std::vector<std::shared_ptr<LinguisticVariable>>,
+                           std::map<std::string, std::shared_ptr<OutputMapper>>,
+                           std::map<std::string, std::shared_ptr<FuzzyControlSystem>>>>();
+
+    return fuzzy_rule_program;
+  } catch (const std::exception &e) {
+    utils::ExceptionHandler::exception("Error while parsing fuzzy rule file: \"{}\". Error: {}", fuzzyRuleFilename,
+                                       e.what());
+    throw std::runtime_error("This should never be reached");
```

Comment:
It avoids compilation warnings, as otherwise, the end of the non-void function could be reached without returning something.

It is redundant as the `utils::ExceptionHandler::exception` function call causes a program-crash anyway, but the compiler can't see this and complains.

I don't know if there are better ways to solve this. Maybe we could leave out the `try-catch` all together and just stick with the default error-messages from ANTLR.

Reply:
Since `utils::ExceptionHandler::exception` is a void function unfortunately not :/

Reply:
Sorry, I mean `throw utils::ExceptionHandler::AutoPasException(....)`.